src/axolotl/state.rs

```diff
@@ -181,72 +180,121 @@ impl <T:Axolotl> ReceiveChain<T> {
-        let receive_chain_position =  self.receive_chains.iter().position(
-            | &ReceiveChain{ref ratchet_key, ..} | axolotl_impl.ratchet_keys_are_equal(ratchet_key, message_ratchet_key.borrow())
+    pub fn decrypt(&mut self, axolotl_impl : &T, message : T::Message, ref mac : T::Mac
+    ) -> Result<T::PlainText,ReceiveError<T>> {
+        let Header{ message_number, message_number_prev, ratchet_key : message_ratchet_key_ref } = try!(
+            axolotl_impl
+                .decode_header(&message)
+                .map_err(|e|{ReceiveError::DecodeError(e)})
         );
+        let message_ratchet_key = message_ratchet_key_ref.borrow();
+
+        self.try_decrypt_with_skipped_chain(axolotl_impl, message, mac, message_number, message_ratchet_key)
+            .or_else(|message| {
+                self.try_decrypt_with_current_chain(axolotl_impl, message, mac, message_number, message_ratchet_key)
+            })
+            .unwrap_or_else(|message| {
+                self.try_decrypt_with_new_chain(axolotl_impl, message_number, message_number_prev, message_ratchet_key, message, mac)
+            })
+    }
 
-        match receive_chain_position {
+    fn try_decrypt_with_skipped_chain(
+        &mut self, 
+        axolotl_impl : &T, 
+        message : T::Message, 
+        mac : &T::Mac, 
+        message_number : usize, 
+        message_ratchet_key : &T::PublicKey
+    ) ->  Result<Result<T::PlainText,ReceiveError<T>>, T::Message> {
+        let position = self.skipped_receive_chains.iter().position(
+            | &ReceiveChain{ref ratchet_key, ..} | axolotl_impl.ratchet_keys_are_equal(ratchet_key, message_ratchet_key)
+        );
+        match position {
             Some(pos) => {
-                let receive_chain = &mut self.receive_chains[pos];
-                receive_chain.try_decrypt(axolotl_impl, message_number, message, mac, &self.identity_key_local, &self.identity_key_remote)
-            }
-            None => {
-                self.try_decrypt_with_new_chain(axolotl_impl, message_number, message_ratchet_key.borrow(), message, mac)
-            }
-                
+                let receive_chain = &mut self.skipped_receive_chains[pos];
+                Ok(receive_chain.try_decrypt(axolotl_impl, message_number, message, mac, &self.session_identity))
+            },
+            None => Err(message),
         }
     }
 
-    fn try_decrypt_with_new_chain<'a>(
+    fn try_decrypt_with_current_chain(
+        &mut self, 
+        axolotl_impl : &T, 
+        message : T::Message, 
+        mac : &T::Mac, 
+        message_number : usize, 
+        message_ratchet_key : &T::PublicKey
+    ) ->  Result<Result<T::PlainText,ReceiveError<T>>, T::Message> {
+        match self.current_receive_chain {
+            Some((ref mut current_chain, ref mut chain_key)) => {
+                if !axolotl_impl.ratchet_keys_are_equal(&current_chain.ratchet_key, message_ratchet_key) {
+                    return Err(message);
```

Comment:
This is some fairly deep nesting, probably worth extracting to a named function so the intention in cleaner